WIP: IPv6 support

guillaumerose opened the pull request:
This PR adds IPv6 support. I still need to refactor some code esp. the 2 hacks needed.

At least, we have something working!

guillaumerose committed:
Add AAAA answers in the built-in DNS server

guillaumerose committed:
Some MAC addresses can be used for multicast. Treat them as broadcast.

It used by NDP protocol.

guillaumerose committed:
Configure the stack to handle IPv6 packets

guillaumerose committed:
Hacks for ipv6

guillaumerose committed:
Add basic e2e tests for ipv6

protosam commented:
https://github.com/containers/gvisor-tap-vsock/blob/3ec335bea2e933a377b4da70d33fb81c63fc07f5/pkg/virtualnetwork/virtualnetwork.go#L100-L110

```go
stack.Options{
		NetworkProtocols: []stack.NetworkProtocolFactory{
			ipv4.NewProtocol,
			ipv6.NewProtocol,
			arp.NewProtocol},
		TransportProtocols: []stack.TransportProtocolFactory{
			tcp.NewProtocol,
			udp.NewProtocol,
			icmp.NewProtocol4,
			icmp.NewProtocol6},
		HandleLocal: false,
	}
```

Not really sure if this is relevant, [it's not even my idea](https://github.com/cloudflare/slirpnetstack/blob/2b45310e2758f6b16b74f521a0232bbff4718715/stack.go#L80-L91). Been trying to understand how this VPN even works and noticed the stack options for slirpnetstack explicitly have some IPv6 conf involved, [but I'm not convinced it even supports IPv6 with this and all the .To4() calls](https://github.com/cloudflare/slirpnetstack/blob/2b45310e2758f6b16b74f521a0232bbff4718715/net.go#L49-L61) (haven't tested).

rhatdan commented:
@guillaumerose Are you still working on this?

guillaumerose commented:
No, not anymore but it was working fine. A user tested it also with success.

rhatdan commented:
Care to update it?
@gbraad @baude PTAL

gbraad commented:
I do not see a compelling benefit for our VM setup.

praveenkumar commented:
It need rebase and also `WIP` so better someone who have bit info around ipv6 test/review it.

protosam commented:
@gbraad the value is that this would just prepare the codebase for the future. ipv4 is a relic of the past that's on life support because it's "good enough" right now while companies struggle to get subnet assignments.

gbraad commented:
I understand, but there is no one on my team who can reserve time for this in the coming weeks. As this is not something we can use now. It will actually adversely affect us.

gbraad commented:
> It need rebase and also `WIP` so better someone who have bit info around ipv6 test/review it.

The 'hacks' remains a commit without clear definition what made them called as such.

protosam commented:
It's definitely nowhere near urgent to get this rolled out. In December, I could put some effort into this. This was something I tried working on, but I ran out of time last year.

gbraad commented:
We prioritized this as part of the crc Podman bundle. Definitely will happen before December. Might ask you for testing/confirmation. After my PTO have to reinstall my router and switch anyway due to a mistake so will make Ipv6 a priority 😝.

protosam commented:
> Might ask you for testing/confirmation.
Definitely happy to help out with that. Even though I left Red Hat earlier this year, I still have a vested personal interest in tools like gvproxy.

gbraad commented:
> I still have a vested personal interest in tools like gvproxy.

What are you using it for?

protosam commented:
Been on using Apple's virtualization framework in place of QEMU for some stability + performance gains to be had with the virtiofs support. Qemu hasn't been very fun since upgrading to the M1 in my experience, but I have a desire to continue using podman over Docker Desktop. 

Also have a code base that I'll be revisiting from last December where I began working on a re-write for gvproxy. It's not exactly secure, because of the access to `192.168.127.1` inside of containers.

gbraad commented:
@protosam We have a driver for this, vfkit that uses vz. However we recently had to fork this driver as the upstream owner is not very helpful: https://github.com/code-ready/crc/issues/3362 it might be an idea to talk to @cfergeau.

protosam commented:
@gbraad Will follow up in that issue. I'm open to contributing on an eco-system around  the vz code base, because I'm already using it.

balajiv113 commented:
Any plans for this feature ??

I tried this patch with latest but facing issues not able to resolve it 😕

cfergeau commented:
Feature is planned, but very low priority at the moment :-/

sanmai-NL commented:
@cfergeau Why is it low priority? IPv6 is very essential and basic on nowadays' Internet?

gbraad commented:
@sanmai-NL The nowadays internet is still mostly IPv4 based. As an example, GitHub does not even have an IPv6 address, so it means you have to run a hybrid stack.

Since we have to prioritize tasks to work on, we have decided first to resolve the virtualization of Podman Machine on macOS and moving to vfkit (instead of Qemu); we experienced issues with qemu+virtio. After this we will first refactor the codebase and improve the buffers. IPv6 will happen at some point, just need to be more strategic. Hope you understand.

Do you have a specific usecase that otherwise will not work?

karta0807913 committed:
Implement DHCPv6

* pkg/tap/ip_pool.go support ipv6 Ip address
* support ipv6 advertisement in switch.go

cfergeau committed:
fixup! Implement DHCPv6, IPv6 DNS AAAA record

cfergeau committed:
fixup! Hacks for ipv6

Adjust to newer gvisor API

Not fully sure about the buffer changes in pkg/tap/router_advertisement.go
This integrates some fixes from Chuxuan Liang PR.

cfergeau committed:
fixup! Configure the stack to handle IPv6 packets

Adjust to newer gvisor API

vyasgun commented:
Tested on macOS with vfkit. Verified IPv6 is working as expected:

```
[fedora@vfkit-vm-2 ~]$ ip -6 addr show dev enp0s1
2: enp0s1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    altname enx5a94efe40cee
    inet6 fd00::2/128 scope global dynamic noprefixroute
       valid_lft 28572sec preferred_lft 14172sec
    inet6 fd00::5894:efff:fee4:cee/64 scope global dynamic noprefixroute
       valid_lft 86171sec preferred_lft 14171sec
    inet6 fe80::5894:efff:fee4:cee/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
[fedora@vfkit-vm-2 ~]$ ip -6 route show default
default via fe80::1 dev enp0s1 proto ra metric 100 pref medium
[fedora@vfkit-vm-2 ~]$ ping6 -c 2 fd00::1
PING fd00::1 (fd00::1) 56 data bytes
64 bytes from fd00::1: icmp_seq=1 ttl=64 time=0.329 ms
64 bytes from fd00::1: icmp_seq=2 ttl=64 time=0.349 ms

--- fd00::1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1036ms
rtt min/avg/max/mdev = 0.329/0.339/0.349/0.010 ms
```

karta0807913 committed:
fix DNS for AAAA records

karta0807913 committed:
remove testing code

Change-Id: Iccbfa76ada834a28e1a40e0c4a0e5b70ec830aae

karta0807913 committed:
fix(lint): expose error from raBuf function

Change-Id: I3f05edf028513329e2f905d18e3ceb290e52c539

cfergeau committed:
Revert "Add basic e2e tests for ipv6"

This reverts commit cd7dceacf0e9890b9cdaec4530829ded3d057035.

These tests currently do not build, we’ll need to make a decision
whether to fix them or drop them.

cfergeau commented:
> P.S. Please help me re-push the branch. I used my work email to commit accidentally.

@karta0807913 This should now be all good, I’ve used "karta0807913" and your gmail address in all commits. Let me know if I missed some places.
I’ve picked up your latest change and added them to this PR, the history is different in this branch and yours, but the content is the same, except for a go module difference which you can solve on your end with `go mod tidy && go mod vendor`.

We’ll need `Signed-off-by` on the commits that you made though, https://github.com/containers/gvisor-tap-vsock/pull/59/checks?check_run_id=62962141326 has some details, can you add this?

karta0807913 commented on pkg/tap/ip_pool.go:
```diff
@@ -47,21 +49,26 @@ func (p *IPPool) GetOrAssign(mac string) (net.IP, error) {
-			continue
+		candidate := net.IP(ipBytes)
+		if !p.base.Contains(candidate) {
+			return nil, errors.New("cannot find available IP")
```
all done. :)

P.S. i cannot push on branch `guillaumerose:ipv62`. so i update on my branch

arikon commented:
@guillaumerose @cfergeau @karta0807913 @vyasgun Thank you for your work on this PR!

I tried to use `gvproxy` from @karta0807913 branch with `podman` on Apple Silicon M3 MacBook Pro, and want to share results with you.

Steps to reproduce:

1. Build `gvproxy` from this repo/branch: https://github.com/karta0807913/gvisor-tap-vsock/tree/ipv6
2. Replace `gvproxy` in `podman` directory, installed using `brew install podman` here: `/opt/homebrew/Cellar/podman/5.7.1/libexec/podman/gvproxy`
3. On `podman` VM (`podman machine ssh`): `tcpdump -ni enp0s1 icmp6`
4. On another ssh session of `podman` VM (`podman machine ssh`): `curl -6 -vk --connect-timeout 5 https://[2a02:6b8:0:3400:0:7ff3:0:2]/` (IP here is IPv6-only host)

So to results:

tcpdump output:

```bash
# sudo tcpdump -ni enp0s1 -vv icmp6
dropped privs to tcpdump
tcpdump: listening on enp0s1, link-type EN10MB (Ethernet), snapshot length 262144 bytes
11:51:49.239918 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:51:50.286313 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:51:51.311262 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:52:50.018201 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:52:51.024624 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:52:52.046314 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:53:53.420391 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fe80::fb99:872d:7a83:dec6 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:54:53.148541 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:54:54.158798 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:54:55.182480 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:55:58.696527 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fe80::fb99:872d:7a83:dec6 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:56:17.490925 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:56:18.510259 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:56:19.536668 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:56:43.562896 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) 2a02:6b8:0:3400:0:7ff3:0:2 > fe80::fb99:872d:7a83:dec6: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::fb99:872d:7a83:dec6
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:dd
	    0x0000:  5a94 efe4 0cdd
11:56:43.563061 IP6 (hlim 255, next-header ICMPv6 (58), payload length 24) fe80::fb99:872d:7a83:dec6 > 2a02:6b8:0:3400:0:7ff3:0:2: [icmp6 sum ok] ICMP6, neighbor advertisement, length 24, tgt is fe80::fb99:872d:7a83:dec6, Flags [solicited]
11:56:49.039219 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fe80::fb99:872d:7a83:dec6 > 2a02:6b8:0:3400:0:7ff3:0:2: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has 2a02:6b8:0:3400:0:7ff3:0:2
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:56:50.063178 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fe80::fb99:872d:7a83:dec6 > 2a02:6b8:0:3400:0:7ff3:0:2: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has 2a02:6b8:0:3400:0:7ff3:0:2
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:56:51.086688 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fe80::fb99:872d:7a83:dec6 > 2a02:6b8:0:3400:0:7ff3:0:2: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has 2a02:6b8:0:3400:0:7ff3:0:2
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:56:56.233742 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:56:57.295801 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:56:58.320148 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fd00::2 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
11:57:59.364807 IP6 (hlim 255, next-header ICMPv6 (58), payload length 32) fe80::fb99:872d:7a83:dec6 > ff02::1:ff00:1: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::1
	  source link-address option (1), length 8 (1): 5a:94:ef:e4:0c:ee
	    0x0000:  5a94 efe4 0cee
^C
23 packets captured
23 packets received by filter
0 packets dropped by kernel
```

curl output:

```bash
# curl -6 -vk --connect-timeout 5 https://[2a02:6b8:0:3400:0:7ff3:0:2]/
*   Trying [2a02:6b8:0:3400:0:7ff3:0:2]:443...
* connect to 2a02:6b8:0:3400:0:7ff3:0:2 port 443 from fd00::2 port 44830 failed: No route to host
* Failed to connect to 2a02:6b8:0:3400:0:7ff3:0:2 port 443 after 3059 ms: Could not connect to server
* closing connection #0
curl: (7) Failed to connect to 2a02:6b8:0:3400:0:7ff3:0:2 port 443 after 3059 ms: Could not connect to server
```

The problem is IPv6-only host is not reacheable from VM.
NDP is not working.

karta0807913 commented:
hi @arikon. thanks for your testing! i made a mistake that we should also accepts the ICMPv6 package from the link local address `fe80::1`. NDP is working on my computer now. please help me test the podman again. thanks!

MrMarvel commented:
Is pr still in progress or stale?

netops2devops commented:
Hey folks, 

thanks for continuing to work on this PR slowly but steadily :) I am on MacOS 26.6 (M4) and just like @arikon I too was able to successfully test IPv6 connectivity from Podman machine using @karta0807913's ipv6 branch. I did the following 

## Build gvproxy from karta0807913's ipv6 branch

```sh
$ git clone git@github.com:karta0807913/gvisor-tap-vsock.git
$ cd gvisor-tap-vsock 
$ git checkout remotes/origin/ipv6
$ make gvproxy
$ cp ./bin/gvproxy /opt/homebrew/Cellar/podman/6.1.0/libexec/podman
```

## Create a new Podman machine and test

```
❯ podman machine init --user-mode-networking

❯ podman machine start
Starting machine "podman-machine-default"


❯ podman machine ssh
Connecting to vm podman-machine-default. To close connection, use `~.` or `exit`
Fedora CoreOS 44.20260720.3.1

core@localhost:~$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
  ... snipped ...

2: enp0s1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 5a:94:ef:e4:0c:ee brd ff:ff:ff:ff:ff:ff
    altname enx5a94efe40cee
    inet 192.168.127.2/24 brd 192.168.127.255 scope global dynamic noprefixroute enp0s1
       valid_lft 3587sec preferred_lft 3587sec
    inet6 fd00::2/128 scope global dynamic noprefixroute
       valid_lft 28790sec preferred_lft 14390sec
    inet6 fd00::cbc1:5d43:11a6:f605/64 scope global dynamic noprefixroute
       valid_lft 86389sec preferred_lft 14389sec
    inet6 fe80::de09:6b:f24d:d197/64 scope link noprefixroute
       valid_lft forever preferred_lft forever

$ curl -6 ifconfig.io
2607:f750:a54:xxxx:xxxx:xxxx:xxxx:xxxx

$ ping6 google.com
PING google.com (2607:f8b0:4009:80d::200e) 56 data bytes
64 bytes from 2607:f8b0:4009:80d::200e: icmp_seq=1 ttl=64 time=0.175 ms
64 bytes from 2607:f8b0:4009:80d::200e: icmp_seq=2 ttl=64 time=0.374 ms
64 bytes from 2607:f8b0:4009:80d::200e: icmp_seq=3 ttl=64 time=0.371 ms
^C
--- google.com ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2085ms
rtt min/avg/max/mdev = 0.175/0.306/0.374/0.093 ms
```

Using `gvproxy` without building from karta0807913's IPv6 branch leads to Podman machine acquiring only IPv4 address. 

@cfergeau @guillaumerose @gbraad Is it possible to proceed with the merging from @karta0807913's ipv6 branch or is there anything else that is needed? Please advise. 

Thanks
